massgov/openmass
# DP-25136 entity usage not showing all relationships

### Description
Entity usage wasn't correctly working when entities were modified or deleted.

### Explanation
Originally when registering usage count on modified entities, the `original` property was present on the modified entity, and the process of updating the usage count needs the `original` property. 

However, because we are using a queue to calculate the usage count later, the `original` property is not on the entity anymore when we load the entity when processing a queue item to calculate usage. The missing `original` property caused the usage count process to skip, misleading results (for editing and deleting).

Existing tests were passing because they were related to the creation of entities, and the process for registering usage count on creation doesn't need the `original` property, hence the usage numbers were correct, and the existing tests were passing.

On this PR, everything is treated as a creation of an entity when registering usage records, avoiding the need for the `original` property on the entity. The entity usage table is cleaned before counting usage by deleting the usage records from older versions of the source entity.

I added tests for the edition and deletion of the source entities.

**Jira:** (Skip unless you are MA staff)
DP-25136


**To Test:**
Before testing, to speed up the usage count and ensure you are processing your recent usage records, **only for development environments**, run the following:

> drush sqlq "TRUNCATE entity_usage" # truncate entity usage table
> drush sqlq "DELETE FROM queue WHERE name='entity_usage_tracker'" # delete entity usage tracking queue

All nodes created for entity usage tests should be published.

---
- [x] Create a node: node1.
- [x] Create a node: node2, and link to node1.
- [x] Run `drush queue:run entity_usage_tracker --time-limit=30  && drush clean_usage_table`
- [x] Go to the second tab "Pages linking here" of node1, notice 1 usage record from node2.
---
- [x] Remove the link on node2.
- [x] Run `drush queue:run entity_usage_tracker --time-limit=30  && drush clean_usage_table`
- [x] Go to the second tab "Pages linking here" of node1, notice no usage records.
---
- [x] Restore the link to node1 on node2.
- [x] Run `drush queue:run entity_usage_tracker --time-limit=30  && drush clean_usage_table`
- [x] Go to the second tab "Pages linking here" of node1, notice 1 usage record from node2.
---
- [x] Delete node2.
- [x] Run `drush queue:run entity_usage_tracker --time-limit=30  && drush clean_usage_table`
- [x] Go to the second tab "Pages linking here" of node1, notice no usage records.

---

[Peer Review Checklist](https://github.com/massgov/openmass/blob/develop/docs/peer_review_checklist.md)
